tracing-subscriber/src/fmt/format/mod.rs

```diff
@@ -1041,40 +1041,79 @@ impl<'a, F> fmt::Debug for FieldFnVisitor<'a, F> {
-pub struct FmtSpan(FmtSpanInner);
+pub struct FmtSpan(u8);
 
 impl FmtSpan {
+    /// one event when span is created
+    pub const NEW: FmtSpan = FmtSpan(1 << 0);
+    /// one event per enter of a span
+    pub const ENTER: FmtSpan = FmtSpan(1 << 1);
+    /// one event per exit of a span
+    pub const EXIT: FmtSpan = FmtSpan(1 << 2);
+    /// one event when the span is dropped
+    pub const CLOSE: FmtSpan = FmtSpan(1 << 3);
+
     /// spans are ignored (this is the default)
-    pub const NONE: FmtSpan = FmtSpan(FmtSpanInner::None);
+    pub const NONE: FmtSpan = FmtSpan(0);
     /// one event per enter/exit of a span
-    pub const ACTIVE: FmtSpan = FmtSpan(FmtSpanInner::Active);
-    /// one event when the span is dropped
-    pub const CLOSE: FmtSpan = FmtSpan(FmtSpanInner::Close);
+    pub const ACTIVE: FmtSpan = FmtSpan(FmtSpan::ENTER.0 | FmtSpan::EXIT.0);
     /// events at all points (new, enter, exit, drop)
-    pub const FULL: FmtSpan = FmtSpan(FmtSpanInner::Full);
+    pub const FULL: FmtSpan =
+        FmtSpan(FmtSpan::NEW.0 | FmtSpan::ENTER.0 | FmtSpan::EXIT.0 | FmtSpan::CLOSE.0);
+
+    /// Check whether or not a certain flag is set for this [`FmtSpan`]
+    fn contains(&self, other: FmtSpan) -> bool {
+        self.clone() & other.clone() == other
+    }
 }
 
+macro_rules! impl_fmt_span_bit_op {
+    ($trait:ident, $func:ident, $op:tt) => {
+        impl std::ops::$trait for FmtSpan {
+            type Output = FmtSpan;
+
+            fn $func(self, rhs: Self) -> Self::Output {
+                FmtSpan(self.0 $op rhs.0)
+            }
+        }
+    };
+}
+
+macro_rules! impl_fmt_span_bit_assign_op {
+    ($trait:ident, $func:ident, $op:tt) => {
+        impl std::ops::$trait for FmtSpan {
+            fn $func(&mut self, rhs: Self) {
+                *self = FmtSpan(self.0 $op rhs.0)
+            }
+        }
+    };
+}
+
+impl_fmt_span_bit_op!(BitAnd, bitand, &);
+impl_fmt_span_bit_op!(BitOr, bitor, |);
+impl_fmt_span_bit_op!(BitXor, bitxor, ^);
+
+impl_fmt_span_bit_assign_op!(BitAndAssign, bitand_assign, &);
+impl_fmt_span_bit_assign_op!(BitOrAssign, bitor_assign, |);
+impl_fmt_span_bit_assign_op!(BitXorAssign, bitxor_assign, ^);
+
 impl Debug for FmtSpan {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self.0 {
-            FmtSpanInner::None => f.write_str("FmtSpan::NONE"),
-            FmtSpanInner::Active => f.write_str("FmtSpan::ACTIVE"),
-            FmtSpanInner::Close => f.write_str("FmtSpan::CLOSE"),
-            FmtSpanInner::Full => f.write_str("FmtSpan::FULL"),
+        if self.contains(FmtSpan::NEW) {
+            f.write_str("FmtSpan::NEW")?;
+        }
+        if self.contains(FmtSpan::ENTER) {
+            f.write_str("FmtSpan::ENTER")?;
+        }
+        if self.contains(FmtSpan::EXIT) {
+            f.write_str("FmtSpan::EXIT")?;
+        }
+        if self.contains(FmtSpan::CLOSE) {
+            f.write_str("FmtSpan::CLOSE")?;
         }
```

Comment:
If multiple bitflags are set, this looks like it will just concatenate them in a kind of ugly way. for example, suppose NEW, ENTER, and EXIT flags are set, this will format that as `FmtSpan::NEWFmtSpan::ENTERFmtSpan::EXIT`. I think we should maybe delimit them with `|` instead, maybe like this:
```suggestion
        let mut wrote_flags = false;
        if self.contains(FmtSpan::NEW) {
            f.write_str("FmtSpan::NEW")?;
            wrote_flags = true;
        }
        
        if self.contains(FmtSpan::ENTER) {
            if wrote_flags {
                f.write_str(" | ")?;
            }
            f.write_str("FmtSpan::ENTER")?;
            wrote_flags = true;
        }
        
        if self.contains(FmtSpan::EXIT) {
            if wrote_flags {
                f.write_str(" | ")?;
            }
            f.write_str("FmtSpan::EXIT")?;
            wrote_flags = true;
        }
        
        if self.contains(FmtSpan::CLOSE) {
            if wrote_flags {
                f.write_str(" | ")?;
            }
            f.write_str("FmtSpan::CLOSE")?;
            wrote_flags = true;
        }
```

we could probably also write a helper function to make that a little less repetitive, if you like...

Reply:
we might also want to add
```rust
if !wrote_flags {
    f.write_str("FmtSpan::NONE")?;
}
```
at the very end, so that an empty set of flags doesn't format itself as the empty string?

Reply:
Whoops, that's a good point, I'll fix it.